local-volume/provisioner/pkg/discovery/discovery.go

```diff
@@ -0,0 +1,116 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/local-volume/provisioner/pkg/common"
+	"github.com/kubernetes-incubator/external-storage/local-volume/provisioner/pkg/types"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/pkg/api/v1"
+)
+
+type Discoverer struct {
+	*types.RuntimeConfig
+}
+
+func NewDiscoverer(config *types.RuntimeConfig) *Discoverer {
+	return &Discoverer{RuntimeConfig: config}
+}
+
+// DiscoverLocalVolumes reads the configured discovery paths, and creates PVs for the new volumes
+func (d *Discoverer) DiscoverLocalVolumes() {
+	for class, path := range d.DiscoveryMap {
+		d.discoverVolumesAtPath(class, path)
+	}
+}
+
+func (d *Discoverer) discoverVolumesAtPath(class, relativePath string) {
+	fullPath := filepath.Join(d.MountDir, relativePath)
+	glog.Infof("Discovering volumes at mount path %q for storage class %q", fullPath, class)
+
+	files, err := d.VolUtil.ReadDir(fullPath)
+	if err != nil {
+		glog.Errorf("Error reading directory: %v\n", err)
+		return
+	}
+
+	for _, file := range files {
+		// Check if PV already exists for it
+		pvName := generatePVName(file, d.NodeName, class)
+		if !d.Cache.PVExists(pvName) {
+			// If not, create PV
+			// TODO: Validate that this path is file-based, and detect capacity
+			d.createPV(file, relativePath, class)
+		}
+	}
+}
+
+// TODO: maybe a better way would be to hash the 3 fields
+func generatePVName(file, node, class string) string {
+	return fmt.Sprintf("%v-%v-%v", class, node, file)
+}
+
+func (d *Discoverer) createPV(file, relativePath, class string) {
+	pvName := generatePVName(file, d.NodeName, class)
+	outsidePath := filepath.Join(d.HostDir, relativePath, file)
+
+	glog.Infof("Found new volume at host path %v, creating LocalVolume PV %q\n", outsidePath, pvName)
+	pvSpec := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: pvName,
+			Annotations: map[string]string{
+				common.AnnProvisionedBy: d.Name,
+			},
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeReclaimPolicy: v1.PersistentVolumeReclaimDelete,
+			Capacity: v1.ResourceList{
+				// TODO: detect capacity
+				v1.ResourceName(v1.ResourceStorage): resource.MustParse("10Gi"),
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+			/** TODO: api not merged yet
+			LocalVolume: &v1.LocalVolumeSource{
+				NodeName: config.NodeName,
+				Fs: v1.LocalFsVolume{Path: outsidePath},
+			},*/
+			},
+			AccessModes: []v1.PersistentVolumeAccessMode{
+				v1.ReadWriteOnce,
```

Comment:
I talked to Saad about how AccessModes work today.  Today, AccessModes is only used to pair PVCs with PVs with the same access modes.  The system does not have any special behavior of enforcing attaches or mounts based on access modes.  Today, it is the underlying storage (like gce pd) that does the "enforcing" of access mode (via attach error). There is a PR out that is looking to enforce attaching based on access modes in Kubernetes itself instead of the underlying storage.

However, there is no plan to support enforcing mounts based on access modes.  The reason for this is because users may have been relying on the current behavior of no mounting enforcement, and we should not break them by adding more strict enforcement.  If people really need mount enforcement, then we would have to provide it through a new field.  Another question is why do you need mount time enforcement though?  You will only share a volume if you specify the same PVC.  And PVCs are namespaced, so that means you are in control of all the pods sharing a PVC, so the enforcement would only be protecting you from yourself.

Considering all of this, AccessMode has no meaning for local volumes, because AccessMode only refers to attaching, and local volumes are already attached.  If we want to enforce some mounting access modes, then we need a new field in the PVC/PV.  It's a big user misconception that access modes refers to mounting, and maybe we need to update the documentation to clarify that.